src/funnels/Depositor.sol

```diff
@@ -187,238 +105,186 @@ contract Depositor {
+
+        if (amt0Owed > 0) GemLike(decoded.gem0).transferFrom(decoded.payer, msg.sender, amt0Owed);
+        if (amt0Owed > 0) GemLike(decoded.gem1).transferFrom(decoded.payer, msg.sender, amt1Owed);
     }
 
-    function _addLiquidity(DepositParams memory p, address to) internal returns (uint128 liquidity, uint256 amt0, uint256 amt1) {
-        bytes32 key = keccak256(abi.encode(p.gem0, p.gem1, p.fee, p.tickLower, p.tickUpper));
-        uint256 tokenId = tokenIds[key];
-        if (tokenId == 0) {
-            PositionManagerLike.MintParams memory params = PositionManagerLike.MintParams({
-                token0: p.gem0,
-                token1: p.gem1,
-                fee: p.fee,
-                tickLower: p.tickLower,
-                tickUpper: p.tickUpper,
-                amount0Desired: p.amt0,
-                amount1Desired: p.amt1,
-                amount0Min: p.minAmt0,
-                amount1Min: p.minAmt1,
-                recipient: to,
-                deadline: block.timestamp
-            });
-            (tokenId, liquidity, amt0, amt1) = PositionManagerLike(uniV3PositionManager).mint(params);
-            tokenIds[key] = tokenId;
-        } else {
-            PositionManagerLike.IncreaseLiquidityParams memory params = PositionManagerLike.IncreaseLiquidityParams({
-                tokenId: tokenId,
-                amount0Desired: p.amt0,
-                amount1Desired: p.amt1,
-                amount0Min: p.minAmt0,
-                amount1Min: p.minAmt1,
-                deadline: block.timestamp
-            });
-            (liquidity, amt0, amt1) = PositionManagerLike(uniV3PositionManager).increaseLiquidity(params);
-        }
+    struct LiquidityParams {
+        address gem0;
+        address gem1;
+        uint24  fee;
+        int24   tickLower;
+        int24   tickUpper;
+        uint128 liquidity;
+        uint256 amt0Desired; // relevant only if liquidity == 0
+        uint256 amt1Desired; // relevant only if liquidity == 0
+        uint256 amt0Min;
+        uint256 amt1Min;
     }
 
-    function deposit(DepositParams memory p) external auth returns (uint128 liquidity, uint256 amt0, uint256 amt1) {
+    function deposit(LiquidityParams memory p)
+        external
+        auth
+        returns (uint128 liquidity, uint256 amt0, uint256 amt1)
+    {
         require(p.gem0 < p.gem1, "Depositor/wrong-gem-order");
 
-        require(block.timestamp >= zzz[p.gem0][p.gem1] + hops[p.gem0][p.gem1], "Depositor/too-soon");
-        zzz[p.gem0][p.gem1] = block.timestamp;
+        Timing memory timing = timings[p.gem0][p.gem1];
+        require(block.timestamp >= timing.zzz + timing.hop, "Depositor/too-soon");
+        timings[p.gem0][p.gem1].zzz = uint64(block.timestamp);
+
+        UniV3PoolLike pool = _getPool(p.gem0, p.gem1, p.fee);
+        liquidity = (p.liquidity == 0)
+            ? _getLiquidityForAmts(pool, p.tickLower, p.tickUpper, p.amt0Desired, p.amt1Desired)
```

Comment:
Feels like it may be cleaner to only have `liquidity` as param in this core `deposit` function and to have another wrapper function `depositAmounts` that uses the desired amounts as params and calls `_getLiquidityForAmts()` followed by `deposit()`. That would maybe require reorganizing the param struct to have a RangeParams including gem0, gem1,  fee, tickLower, tickUpper and the other liquidity/amount params passed individually.

Reply:
as discussed we will explore in a separate PR.